tools/build-tool/data/android/integrated-services/admob/app/src/main/java/net/sourceforge/castleengine/ServiceAdMob.java

```diff
@@ -42,37 +56,181 @@ private void initialize(String bannerUnitId, String interstitialUnitId, String[]
     }
 
     private void interstitialInitialize()
     {
+        if (mInterstitialUnitId == "") 
+            return;
+
         // Create the interstitial.
         interstitial = new InterstitialAd(getActivity());
         interstitial.setAdUnitId(mInterstitialUnitId);
         // Set an AdListener.
         interstitial.setAdListener(new AdListener() {
-            // unused now
-            // @Override
-            // public void onAdLoaded() {
-            // }
+            @Override
+            public void onAdLoaded() {
+                logInfo(CATEGORY, "onAdLoaded");
+                failedToLoadInterstitialLastTime = false;
+                if (interstitialOpenWhenLoaded) {
+                    interstitialOpenWhenLoaded = false;
+                    logInfo(CATEGORY, "Show ad after waiting for add.");
+                    interstitial.show();
+                }
+
+            }
+
+            @Override
+            public void onAdFailedToLoad(int errorCode)  {
+                logInfo(CATEGORY, "onAdFailedToLoad");
+                failedToLoadInterstitialLastTime = true;
+                if (interstitialOpenWhenLoaded) {
+                    switch(errorCode) {
+                        case AdRequest.ERROR_CODE_INTERNAL_ERROR:
+                            fullScreenAdClosed(TAdWatchStatus.wsUnknownError);
+                            break;
+                        case AdRequest.ERROR_CODE_INVALID_REQUEST:
+                            fullScreenAdClosed(TAdWatchStatus.wsInvalidRequest);
+                            break;
+                        case AdRequest.ERROR_CODE_NETWORK_ERROR:
+                            fullScreenAdClosed(TAdWatchStatus.wsNetworkNotAvailable);
+                            break;
+                        case AdRequest.ERROR_CODE_NO_FILL:
+                            fullScreenAdClosed(TAdWatchStatus.wsNoAdsAvailable);
+                            break;
+                        default:
+                            fullScreenAdClosed(TAdWatchStatus.wsUnknownError);
+                    }
+                }
+                interstitialOpenWhenLoaded = false;
+            }
 
             @Override
             public void onAdClosed() {
                 logInfo(CATEGORY, "Ad Closed");
-                fullScreenAdClosed(true);
+                fullScreenAdClosed(TAdWatchStatus.wsWatched);
+                loadInterstitial(); // load next ad
             }
         });
 
         // Begin loading your interstitial.
-        interstitial.loadAd(buildAdRequest());
+        loadInterstitial();
+    }
+
+    public void loadInterstitial() {
+        failedToLoadInterstitialLastTime = false;
+        if (!interstitial.isLoaded()) {
+            logInfo(CATEGORY, "start loading interstitial");
+            interstitial.loadAd(buildAdRequest());
+        }
+    }
+
+
+    private void rewardedInitialize()
+    {
+        if (mRewardedUnitId == "")
```

Comment:
This should use `equals` method, like `mRewardedUnitId .equals("")`, see my other comment.